python/ray/data/_internal/split.py

```diff
@@ -110,16 +111,33 @@ def _split_single_block(
             input_files=meta.input_files,
             exec_stats=stats.build(),
         )
-        split_result.append((ray.put(split_block), split_meta))
+        split_result.append((ray.put(split_block), split_meta, accessor.num_rows()))
         prev_index = index
     return (block_id, split_result)
 
 
+def _drop_empty_block_split(block_split_indices: List[int], num_rows: int) -> List[int]:
```

Comment:
This seems to have implication on public API Dataset.split_at_indices? It currently can generate empty blocks but this will change it.

Reply:
How about dropping only the ones for start/end of block?

Reply:
I think I'm fine with completely dropping empty blocks (unless they have a utility that I'm not aware of). In that case, we can just go to Dataset.split_at_indices and revamp the semantics there (we just need to de-dup the global indices and document the API that it will do so).

Reply:
@jianoaix ah this is internal change but not change the split_at_indices public api. So what happens is previously we will have dataset who contains empty block (block with 0 rows), now we remove the block but the number of (split) dataset is still the same.

Reply:
Yeah this PR does not change `split_at_indices` semantics.

Wondering would it be better just embed this function's logic inside `_generate_per_block_split_indices()` L70-72 ? We can just avoid adding these unuseful indices in the first place.